### PR TITLE
Entity Lifecycle Stress Tests

### PR DESCRIPTION
**Pull request by simpsont-oci**
Tests to stress rapid creation / destruction of DDS entities using short-lived test processes (publishers and subscribers).
Number of processes and transports are relatively easy to configure.

### CODE REVIEW

File: tests/DCPS/EntityLifecycleStress/EntityLifeCycleStress.mpc [hunk 1]
```diff
@@ -0,0 +1,21 @@
+project(DDS*Publisher) : dcpsexe, dcps_test, dcps_cm, dcps_transports_for_test {
+  exename = publisher
+
+  Idl_Files {
+  }
+
```
Comment (mitza-oci) [inline tests/DCPS/EntityLifecycleStress/EntityLifeCycleStress.mpc:6]:
```suggestion
```

File: tests/DCPS/EntityLifecycleStress/publisher.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        ACE_OS::sleep(delay);
+        ++message.count;
+        message_dw->write(message, handle);
+      }
+
+      {
+        stringstream ss;
+        ss << "Publisher " << ACE_OS::getpid() << " is done. Exiting." << endl;
+
+        cout << ss.str() << flush;
```
Comment (mitza-oci) [inline tests/DCPS/EntityLifecycleStress/publisher.cpp:114]: Why does this need a separate `stringstream`?

File: tests/DCPS/EntityLifecycleStress/publisher.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        message_dw = 0;
+
+        pub->delete_datawriter(dw);
+        participant->delete_publisher(pub);
+      }
+
+      participant->delete_contained_entities();
+      dpf->delete_participant(participant.in());
+  }
+  catch (CORBA::Exception& e)
```
Comment (mitza-oci) [inline tests/DCPS/EntityLifecycleStress/publisher.cpp:128]:
```suggestion
  catch (const CORBA::Exception& e)
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: tests/DCPS/EntityLifecycleStress/publisher.cpp (hunk 2 above)
Comment (simpsont-oci) [reply tests/DCPS/EntityLifecycleStress/publisher.cpp]: I can't remember if it's only a problem on some platforms and not others, but multi-process IO to the console can potentially switch processes between << operators unless it's done as a single output stream. So for the output of these tests I tried to protect against it with stringstream.

**Review by mitza-oci — APPROVED** [review]

File: tests/DCPS/EntityLifecycleStress/publisher.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
+#include "MessengerTypeSupportImpl.h"
+
+#include <dds/DCPS/Service_Participant.h>
+#include <dds/DCPS/Marked_Default_Qos.h>
+#include <dds/DCPS/PublisherImpl.h>
+#include <dds/DCPS/Qos_Helper.h>
+#include "dds/DCPS/StaticIncludes.h"
+#include "dds/DCPS/unique_ptr.h"
```
Comment (iguessthislldo) [inline tests/DCPS/EntityLifecycleStress/publisher.cpp:8]:
```suggestion
#include <dds/DCPS/StaticIncludes.h>
#include <dds/DCPS/unique_ptr.h>
```

File: tests/DCPS/EntityLifecycleStress/publisher.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+#include "tests/Utils/StatusMatching.h"
+
+#ifdef ACE_AS_STATIC_LIBS
+#include <dds/DCPS/RTPS/RtpsDiscovery.h>
+#include <dds/DCPS/transport/rtps_udp/RtpsUdp.h>
+#endif
+
+#include <ace/streams.h>
+#include "ace/Get_Opt.h"
+#include "ace/OS_NS_unistd.h"
```
Comment (iguessthislldo) [inline tests/DCPS/EntityLifecycleStress/publisher.cpp:20]:
```suggestion
#include <ace/Get_Opt.h>
#include <ace/OS_NS_unistd.h>
```

File: tests/DCPS/EntityLifecycleStress/publisher.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+#include <memory>
+
+using namespace Messenger;
+using namespace std;
+
+int ACE_TMAIN(int argc, ACE_TCHAR *argv[]) {
```
Comment (iguessthislldo) [inline tests/DCPS/EntityLifecycleStress/publisher.cpp:27]:
```suggestion
int ACE_TMAIN(int argc, ACE_TCHAR *argv[])
{
```

File: tests/DCPS/EntityLifecycleStress/subscriber.cpp [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+#ifdef ACE_HAS_CPP11
+    std::chrono::milliseconds ms((deadline - OpenDDS::DCPS::MonotonicTimePoint::now()).value().get_msec());
+    std::unique_lock<std::mutex> lock(mutex_);
+#else
+    ACE_Guard<ACE_Thread_Mutex> g(mutex_);
+#endif
+#ifdef ACE_HAS_CPP11
+    while (!valid_data_seen && cv_.wait_for(lock, ms) != cv_status::timeout) {
+#else
+    while (!valid_data_seen && !(cv_.wait(&deadline.value()) == -1 && errno == ETIME)) {
```
Comment (iguessthislldo) [inline tests/DCPS/EntityLifecycleStress/subscriber.cpp:103]: `ACE_Condition` requires configuration to use monotonic time or else it thinks it's being given system time. See `DCPS::ConditionVariable`.

File: tests/DCPS/EntityLifecycleStress/subscriber.cpp [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+  std::mutex mutex_;
+  std::condition_variable cv_;
+#else
+  ACE_Thread_Mutex mutex_;
+  ACE_Condition<ACE_Thread_Mutex> cv_;
+#endif
+  bool valid_data_seen;
+};
+
+int ACE_TMAIN(int argc, ACE_TCHAR *argv[]){
```
Comment (iguessthislldo) [inline tests/DCPS/EntityLifecycleStress/subscriber.cpp:118]:
```suggestion
int ACE_TMAIN(int argc, ACE_TCHAR *argv[])
{
```

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

**Pull request merged**